lxd/auth/driver_tls.go

```diff
@@ -1,65 +1,176 @@
+	}
+
+	authenticationProtocol := details.authenticationProtocol()
+	if authenticationProtocol != api.AuthenticationMethodTLS {
+		t.logger.Warn("Authentication protocol is not compatible with authorization driver", logger.Ctx{"protocol": authenticationProtocol})
+		// Return nil. If the server has been configured with an authentication method but no associated authorization driver,
+		// the default is to give these authenticated users admin privileges.
+		return nil
+	}
+
+	certType, isNotRestricted, projectNames, err := t.certificateDetails(details.username())
+	if err != nil {
+		return err
+	}
+
+	if isNotRestricted || (certType == certificate.TypeMetrics && entitlement == EntitlementCanViewMetrics) {
+		return nil
+	}
+
+	if details.isAllProjectsRequest {
+		// Only admins (users with non-restricted certs) can use the all-projects parameter.
+		return api.StatusErrorf(http.StatusForbidden, "Certificate is restricted")
+	}
+
+	// Check server level object types
+	switch object.Type() {
+	case ObjectTypeServer:
+		if entitlement == EntitlementCanView || entitlement == EntitlementCanViewResources || entitlement == EntitlementCanViewMetrics {
+			return nil
+		}
+
+		return api.StatusErrorf(http.StatusForbidden, "Certificate is restricted")
+	case ObjectTypeStoragePool, ObjectTypeCertificate:
+		if entitlement == EntitlementCanView {
+			return nil
+		}
+
+		return api.StatusErrorf(http.StatusForbidden, "Certificate is restricted")
+	}
+
+	// Check project level permissions against the certificates project list.
+	projectName := object.Project()
+	if !shared.ValueInSlice(projectName, projectNames) {
+		return api.StatusErrorf(http.StatusForbidden, "User does not have permission for project %q", projectName)
+	}
 
-// RenameProject is a no-op. It notifies the authorization service that a project has been renamed.
-func (a *tls) RenameProject(projectID int64, newName string) error {
 	return nil
 }
 
-// StopStatusCheck is a no-op.
-func (a *tls) StopStatusCheck() {
-}
+// GetPermissionChecker returns a function that can be used to check whether a user has the required entitlement on an authorization object.
+func (t *tls) GetPermissionChecker(ctx context.Context, r *http.Request, entitlement Entitlement, objectType ObjectType) (PermissionChecker, error) {
+	allowFunc := func(b bool) func(Object) bool {
+		return func(Object) bool {
+			return b
+		}
+	}
 
-func (a *tls) UserAccess(username string) (*UserAccess, error) {
-	return &UserAccess{Admin: true}, nil
-}
+	details, err := t.requestDetails(r)
+	if err != nil {
+		return nil, api.StatusErrorf(http.StatusForbidden, "Failed to extract request details: %v", err)
+	}
+
+	if details.isInternalOrUnix() {
+		return allowFunc(true), nil
+	}
+
+	authenticationProtocol := details.authenticationProtocol()
+	if authenticationProtocol != api.AuthenticationMethodTLS {
+		t.logger.Warn("Authentication protocol is not compatible with authorization driver", logger.Ctx{"protocol": authenticationProtocol})
```

Comment:
I wonder if we should ban all requests that use an invalid authentication protocol rather than fail open?

Reply:
I wasn't sure. Currently if you authenticate with Candid (without RBAC) or OIDC the default is to allow everything. I can change it but will need to change some tests. Any existing users using these authentication methods will find that they are unauthorized and will need to configure authorization via the unix socket.

Reply:
I think it's unlikely that there many (or any) users using OIDC or standalone Candid though.

Reply:
I see, so they dont have a trusted client TLS cert because they authenticated through a different mechanism but the active authorizer is for TLS. I think we should keep it the same, but perhaps clarify that section there with a comment how you explained it above.